### ISSUE TITLE
Bug: XSRF through sesskey

### ISSUE THREAD

**Issue by skodak** [michael-milette/moodle-filter_filtercodes#246]
### Prerequisites

- [X] I'm reporting the issue to the correct repository.
- [X] I am running the latest version of this plugin (if not, please upgrade first).
- [X] I checked the [README.md](../blob/master/README.md) and [Wiki](../wiki) documentation but did not find my answer.
- [X] I searched to make sure that this issue had not already been reported.
- [ ] I have, or I know someone who has, the ability and willingness to test the fix on a Moodle LMS-based site.
- [X] I am submitting no more than one bug report per issue ticket (please do not submit multiple bugs in one issue ticket).
- [X] I understand that support is provided by volunteer(s).

### What happened?

sesskey secret must not be added through filters for obvious security reasons

### What do you think should have happened?

I expected that the {sesskey} would be removed when the security problem was reported, but it was not. That is why I am including exploit example here.

### Steps to reproduce

1. Login as 'student'
2. Go to 'any forum where you can post'
3. Click on 'Reply'
4. Switch to html source code view
5. Add this html markup (change URL if needed, id is your own account id): `<img src="/admin/roles/admins.php?confirmadd=3&amp;sesskey=%7Bsesskey%7D" alt="&nbsp;" />`
6. Wait for admin to see the post while logged in
7. Login and enjoy ownership of a new Moodle site where you are now a new admin

### Moodle version

all versions

### Installed Moodle UI Language packs.

- [X] English only
- [ ] One or more other languages/regional dialects (specify below)
- [ ] Other

### Workaround

Delete /filter/filtercodes/ directory.

### Additional information

I am providing the same exploit that led to introduction of sesskey in Moodle, hopefully this will finally persuade the author to fix this security hole.

By the way I have found a number of other security problems of different types in this plugin, so I would strongly recommend to advice customers to uninstall the plugin and wait until it gets redesigned with security in mind.

### System information

Client Device type: any
Client Operating System: any
Client Web Browser: any
Server Operating System: any
Server Webserver: any
Server PHP: any

### Code of Conduct

- [X] I have read and agree to the project's <a href="../blob/master/CODE_OF_CONDUCT.md" target="_blank">code of conduct</a>.

**Comment by michael-milette** [comment]
Peter, while I appreciate your intention, you are exhibiting unprofessional behaviour, putting site owners in danger and attempting to coerce. This is a direct violation of the repositories code of conduct that you agreed to. If you continue with this unprofessional behaviour, I will have no other choice than to ban you from this repo. You will not receive any additional warnings.

Did you try the latest release of FilterCodes?

**Comment by skodak** [comment]
@andrewhancox I would recommend you read more about XSRF and XSS, in short there is no way for student to add anything from M.cfg object to any url in forum post because all texts from students are sanitised with HTMLPurifier.

Good resource to learn about web security is OWASP at https://owasp.org/www-community/attacks/csrf

**Comment by andrewhancox** [comment]
Totally down with the OWASP top 10 - I get the issue now, the filter will enable users to build a URL in a post with a valid sesskey that an instructor could inadvertently open.
I was just looking at it from the perspective XSRF/XSS attacks from other applications...

**Comment by michael-milette** [comment]
@skodak thank you for your comments. I try to see past your poor attitude as the points you bring up are actually very helpful use cases that I had not considered. I will be taking these into consideration.

Please keep in mind that I do this on a voluntary basis. Nobody is paying me to work on this plugin. I am sorry that resolving this issue isn't moving ahead as quickly a you might prefer. Consider financially sponsoring the development of solutions to these issues if you would like it to move faster.

Is there a reason that you keep recommending that people uninstall the plugin rather than just disable the sesskey tag in the settings or make use of the Word Censorship plugin to disable any tags they don’t want?

**Comment by skodak** [comment]
@michael-milette I do not like you blaming me for you incompetence in relation to security of web applications, I do not want to be involved with you

**Comment by skodak** [comment]
Here are all the previous reports I found that correctly state CSRF vulnerability through {sesskey} tag:

* https://github.com/michael-milette/moodle-filter_filtercodes/issues/227
* https://github.com/michael-milette/moodle-filter_filtercodes/issues/150
* https://github.com/michael-milette/moodle-filter_filtercodes/pull/130

There was no information where to report security issues when I posted the proof of vulnerability, I did not break any rules of responsible disclosure at that time, everybody could read about this vulnerability in your github issues since 2020. You cannot say it was a real exploit because the {sesskey} is still marketed as a valuable "feature". It was meant to show everybody that it is a very serious security problem that is extremely easy to exploit (that is the nature of CSRF after all).

If I reported the vulnerability without proof nothing would happen, the proof is that even after I published the demonstration image tag the vulnrability WAS NOT FIXED.

Backdating your security policies to make me look bad with this message "UPDATE: The message was deleted for failure to conform with the repository's code of conduct." is not a professional thing to do.

Why was I called unprofessional here? Was it because I posted the ultimate proof that the plugin is vulnerable? Which you ignored again. Was it because I made a reference to _The Emperor’s New Clothes_ folktale? I apologise for that.
